### ISSUE TITLE
Efficiently handling lots of multiselects on one page

### ISSUE THREAD

**Issue by BrendaH** [ehynds/jquery-ui-multiselect-widget#765]
**Usecase:**
I have an interface which has about 100 multiselects and each multiselect has about 120 options. As you can guess this takes quite a while to be converted to the pretty multiselects from this plugin (about 6 seconds).

**Suggested solution:**
In my own copy of the code (version 2) I made some adjustments to have the options only drawn when needed (on opening a multiselect I call refresh and I prevent all options from being drawn on the first init).
I'd be willing to make this into an option and make a pull request, but I'm not sure how to go about this.

A part of the 'solution' requires [the update function](https://github.com/ehynds/jquery-ui-multiselect-widget/blob/c7fbab3530ff3a441ec22dd5b732e1b2c08a66be/src/jquery.multiselect.js#L281-L305) to change: the button should use the text from the actual option tags and not the text that is next to the rendered checkboxes (since they won't be there on init, but you do want to already show the buttons). This needs altering of these 3 lines of code.
```
      var $inputs = this.inputs;
      var $checked = $inputs.filter(':checked');
[snip]
          value = $checked.map(function() { return $(this).next().text(); }).get().join(o.selectedListSeparator);
```
To: 
```
      var $inputs = this.element.find('option');
      var $checked = $inputs.filter(':selected');
[snip]
          value = $checked.map(function() { return $(this).text(); }).get().join(o.selectedListSeparator);
```
The other part has two possible solutions:

1. A new option called lazyDrawOptions. Inside the init function the _buildOptionList would be skipped if this option is true and in the open function _buildOptionList will be triggered (so only for the one being opened at that moment) with a check that it hasn't been drawn already. So each checkbox-list is only drawn once, on first opening of the multiselect.
2. The other possibility is to have an option more like a refreshOnOpen. This also skips the __buildOptionList on init but then just calls refresh on open (every time). This also allows a user to edit the actual select but have the list of checkboxes always be up to date on open. For me personally this is useful because I do some editing (css styles, text) on the checkboxes afterwards and refreshing on open makes it unnecessary to undo those changes again.

This building options/calling refresh inside [the open function](https://github.com/ehynds/jquery-ui-multiselect-widget/blob/c7fbab3530ff3a441ec22dd5b732e1b2c08a66be/src/jquery.multiselect.js#L727) would have to take place before [the check on beforeopen etc.](https://github.com/ehynds/jquery-ui-multiselect-widget/blob/c7fbab3530ff3a441ec22dd5b732e1b2c08a66be/src/jquery.multiselect.js#L735-L738) (otherwise beforeopen can not do anything with the checkboxes). So it might be a bit 'heavy' and redundant in case the if statement evaluates to true and the open function immediately returns.

This is all talking about version 2 code, by the way. I've taken a look at version 3 and did not see an option that takes care of this. So another question would be if I'd implement this in the version 3 code or in version 2.

So:
1 - Is altering the button update function as proposed acceptable?
2 - Which of the proposed options would be better (lazyDrawOptions or refreshOnOpen)?
3 - In which version of the code (2 / 3) should this be implemented?

**Comment by mlh758** [comment]
You should really try out version 3. @SteveTheTechie already made substantial improvements to the load times of the widgets. A 5000 item list now loads in ~0.3 seconds compared to ~2.5 seconds.

**Comment by mlh758** [comment]
If you still have a need for it with version 3 then the change can go into that branch. I would prefer to have version 2 changes be limited to bug fixes only at this point.

**Comment by SteveTheTechie** [comment]
> You should really try out version 3. @SteveTheTechie already made substantial improvements to the load times of the widgets. A 5000 item list now loads in ~0.3 seconds compared to ~2.5 seconds.

Yeah, the reason I did that was because I had a similar use case and was also bumping into the performance issues.  I had even tried loading asynchronously (my tweak).

If you do the option building in the open(), aren't you just moving the time from one place to another?  You would certainly get the widget to load even faster, but then your users have to wait for the menu to open.

Don't get me wrong, I do find this intriguing, but I would also like to see an infinity scrolling approach implemented at some point which would also speed up the open and kind of render all this moot, since it only renders what you see in the div plus some small amount for smooth scrolling.  Everything else is just an empty li to maintain position on the scroll bar.

**Comment by SteveTheTechie** [comment]
The following is problematic , because when options.selectedList is set to function, the function has historically expected to get input elements not option elements.    This means that any code that uses a function for selectedList would be broken, because inputs use the checked property and options use the selected property.

```
  var $inputs = this.element.find('option');
      var $checked = $inputs.filter(':selected');
[snip]
          value = $checked.map(function() { return $(this).text(); }).get().join(o.selectedListSeparator);
```

**Comment by SteveTheTechie** [comment]
The simple way to do this is just check if `this.$checkboxes` exists.   If they exist, do nothing.  If they do not exist, then build them.

**Comment by mlh758** [comment]
Before we go trying to change anything I'd like some feedback from @BrendaH to see if the current improvements are enough.

**Comment by BrendaH** [comment]
Sorry, I got caught up in other work. But I just did a test and version 3 is fast enough for me without the need to postpone the conversion of options to checkboxes.
So the adjustments that I proposed above aren't necessary (at least for me) for version 3.

There's just one thing. Perhaps I am missing something, but the filter interface does not work. I get a filter input, but when I enter anything it has no effect whatsoever. Did anything change in the way the filter should be called or used? I still do:
`$("table#formset select").multiselect(multiselectoptions).multiselectfilter();`
And I do get the filter interface, but it does not work.
